### PR TITLE
Available CPU predictions

### PR DESCRIPTION
**Pull request by alejandrods**
- Modify filters.py for CPU predictions
- Modify visualizate.py for CPU predictions
- Modify loss.py for CPU predictions
- Update Notebooks for using CPU predictions
- Update Readme

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -6,3 +6,5 @@ youtube-dl>=2019.4.17
 jupyterlab
 opencv-python>=3.3.0.10
 pillow
+flask
```
Comment (alexandrevicenzi) [inline requirements.txt]: This should not be here, requirements.txt is only for Jupyter related dependencies.

File: fasterai/loss.py [hunk 2]
```diff
@@ -1,15 +1,25 @@
+import os
 from fastai import *
 from fastai.core import *
 from fastai.torch_core import *
 from fastai.callbacks  import hook_outputs
 import torchvision.models as models
 
 
+cuda = True
```
Comment (alexandrevicenzi) [inline fasterai/loss.py]: You can remote this and use `torch.cuda.is_available()`.

File: fasterai/loss.py [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
+if os.environ['CUDA_VISIBLE_DEVICES'] == '':
+    cuda = False
+
+
 class FeatureLoss(nn.Module):
     def __init__(self, layer_wgts=[20,70,10]):
         super().__init__()
 
-        self.m_feat = models.vgg16_bn(True).features.cuda().eval()
+        if cuda:
```
Comment (alexandrevicenzi) [inline fasterai/loss.py]: Use `torch.cuda.is_available()`.

File: fasterai/visualize.py [hunk 4]
```diff
@@ -11,22 +11,31 @@
...
 class ModelImageVisualizer():
     def __init__(self, filter:IFilter, results_dir:str=None):
         self.filter = filter
         self.results_dir=None if results_dir is None else Path(results_dir)
         self.results_dir.mkdir(parents=True, exist_ok=True)
     
     def _clean_mem(self):
-        torch.cuda.empty_cache()
-        #gc.collect()
+        if cuda:
```
Comment (alexandrevicenzi) [inline fasterai/visualize.py]: Use `torch.cuda.is_available()`.

File: fasterai/filters.py [hunk 5]
```diff
@@ -42,8 +48,16 @@ def _model_process(self, orig:PilImage, sz:int)->PilImage:
         x =  pil2tensor(model_image,np.float32)
         x.div_(255)
         x,y = self.norm((x,x), do_x=True)
-        result = self.learn.pred_batch(ds_type=DatasetType.Valid, 
-            batch=(x[None].cuda(),y[None]), reconstruct=True)
+
```
Comment (alexandrevicenzi) [inline fasterai/filters.py]: Use `torch.cuda.is_available()`.

File: app.py [hunk 6]
```diff
@@ -1,40 +1,36 @@
...
-from os import path
 import torch
 
-import fastai
-from fasterai.visualize import *
-from pathlib import Path
-import traceback
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 
+if device == 'cuda':
```
Comment (alexandrevicenzi) [inline app.py]: Use `torch.cuda.is_available()`.

File: app-video.py [hunk 7]
```diff
@@ -1,40 +1,33 @@
...
 import torch
 
-import fastai
-from fasterai.visualize import *
-from pathlib import Path
-import traceback
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 
+if device == 'cuda':
+    os.environ['CUDA_VISIBLE_DEVICES'] = '0'
```
Comment (alexandrevicenzi) [inline app-video.py]: Use `torch.cuda.is_available()` only, why set a variable?

**Comment by alexandrevicenzi** [comment]
You also need to rebase you patch and fix all conflicts.

File: fasterai/filters.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 from fastai import *
 import math
 from scipy import misc
+import os
```
Comment (alexandrevicenzi) [inline fasterai/filters.py]: Import not used.

File: fasterai/filters.py [hunk 9]
```diff
@@ -42,8 +43,16 @@ def _model_process(self, orig:PilImage, sz:int)->PilImage:
         x =  pil2tensor(model_image,np.float32)
         x.div_(255)
         x,y = self.norm((x,x), do_x=True)
-        result = self.learn.pred_batch(ds_type=DatasetType.Valid, 
-            batch=(x[None].cuda(),y[None]), reconstruct=True)
+
+        if torch.cuda.is_available():
```
Comment (alexandrevicenzi) [inline fasterai/filters.py]: You could refactor this to avoid this duplicated code, the only thing that needs to be in the if is the batch creation.

File: fasterai/loss.py [hunk 10]
```diff
@@ -1,3 +1,4 @@
+import os
```
Comment (alexandrevicenzi) [inline fasterai/loss.py]: Import not used.
